File: 0020_Valid_Parentheses/Solution2.py
class Solution:
    def isValid(self, s: str) -> bool:
        stk = []
        for ch in s:
            if ch in ['(', '[', '{']:
                stk.append(ch)
            else:
                if not stk:
                    return False
                last = stk.pop()
                if ch == ')' and last != '(':
                    return False
                if ch == ']' and last != '[':
                    return False
                if ch == '}' and last != '{':
                    return False
        return stk == []


def test(test_name, s, expected):
    res = Solution().isValid(s)
    if res == expected:
        print(test_name + ' success.')
    else:
        print(test_name + ' failed.')


if __name__ == "__main__":
    s1 = "()"
    expected1 = True
    test("test1", s1, expected1)

    s2 = "()[]{}"
    expected2 = True
    test("test2", s2, expected2)

    s3 = "(]"
    expected3 = False
    test("test3", s3, expected3)

    s4 = "([)]"
    expected4 = False
    test("test4", s4, expected4)

    s5 = "{[]}"
    expected5 = True
    test("test5", s5, expected5)

    s6 = "(])"
    expected6 = False
    test("test6", s6, expected6)

File: 0069_Sqrt(x)/Solution2.cpp
#include <iostream>
using namespace std;

class Solution {
public:
    int mySqrt(int x) {
        if (x == 0)
            return 0;
        
        long long new_width = x;    // 用long long 不然会溢出
        while (x < new_width * new_width) {
            new_width = (new_width + x / new_width) / 2;
        }
        return new_width;
    }
};


void test(string test_name, int x, int expected)
{
    int res = Solution().mySqrt(x);
    if (res == expected) {
        cout << test_name << " success." << endl;
    } else {
        cout << test_name << " failed." << endl;
    }
}

int main()
{
    int x1 = 4;
    int expected1 = 2;
    test("test1", x1, expected1);

    int x2 = 8;
    int expected2 = 2;
    test("test2", x2, expected2);

    int x3 = 2147395600;
    int expected3 = 46340;
    test("test3", x3, expected3);

    return 0;
}
